crates/parser/src/grammar/items/traits.rs

```diff
@@ -50,9 +50,17 @@ pub(super) fn impl_(p: &mut Parser) {
     // test impl_def_neg
     // impl !Send for X {}
     p.eat(T![!]);
-    impl_type(p);
+    let ty_first = impl_type(p);
     if p.eat(T![for]) {
         impl_type(p);
+        if let Some(marker) = ty_first {
+            if marker.kind() != PATH_TYPE {
+                // test_err impl_non_path_type_for_type
+                // impl &T for X {}
+                p.error("expected a trait, found a type");
+                marker.undo_completion(p).complete(p, ERROR);
+            }
+        }
```

Comment:
To make parsing work, all the type parsing functions return a `CompletedMarker` now so that `type_` may return it, is this the proper way? Couldn't think of anything else.

Also the remapping of the marker type to error doesn't work as I had hoped, see the generated RAST, as it remaps the `ty for ty` part to error instead of just the first ty. Really uncertain about the parsing changes in general

Reply:
I guess another Option would be to not change anything here at all, and instead do a check in the validation layer as well as changing the `trait_` method on `ast::Impl` to only return for `PathType`s as it does with this PR?